src/workerd/api/http.h

```diff
@@ -694,9 +694,10 @@ struct RequestInitializerDict {
   jsg::WontImplement credentials;
 
   // In browsers this controls the local browser cache. For Cloudflare Workers it could control the
-  // Cloudflare edge cache. Note that this setting is different from using the `Cache-Control`
-  // header since `Cache-Control` would be forwarded to the origin.
-  jsg::Unimplemented cache;
+  // Cloudflare edge cache. While the standard defines a number of values for this property, our
+  // implementation supports only three: undefined (identifying the default caching behavior that
+  // has been implemented by the runtime), "no-store", and "no-cache".
+  jsg::Optional<kj::String> cache;
```

Comment:
I think we have to guard this type change with the compat flag.

But I see now the issue you have with this, which is that JSG_STRUCT doesn't currently support making properties conditional on compat flag. It may be necessary to extend the JSG_STRUCT macro somehow, or allow multiple JSG_STRUCT declarations where one is chosen based on flags...

Reply:
Is it not sufficient to error on construction?

Reply:
Perhaps also we could have another type that is WasUnimplemented, that checks the compat flag upon construction?

Reply:
> Is it not sufficient to error on construction?

Hmm. I guess it could be, as long as:
* Every function which accepts RequestInit performs the check immediately.
* We throw _exactly_ the same error that we have in the past. This may sound a little extreme, but given we've already broken people twice by doing things that we thought wouldn't break anyone, I think we have to be extreme in our caution at this point.

Here's an idea: What if we extend JSG_STRUCT so that you can give the struct a `validate()` method, which is automatically called after construction, and has the opportunity to throw an exception. That can be used to confidently solve the first point above.

Reply:
> > Is it not sufficient to error on construction?
> 
> Hmm. I guess it could be, as long as:
> 
> * Every function which accepts RequestInit performs the check immediately.
> * We throw _exactly_ the same error that we have in the past. This may sound a little extreme, but given we've already broken people twice by doing things that we thought wouldn't break anyone, I think we have to be extreme in our caution at this point.

So after looking through this I somewhat disagree with my earlier statement because in the future if we decide to implement some behavior that was previously `jsg::Unimplemented` we want it to be intuitive to gate just that field/function.


> Here's an idea: What if we extend JSG_STRUCT so that you can give the struct a validate() method, which is > automatically called after construction, and has the opportunity to throw an exception. That can be used to confidently solve the first point above.

While this method seems good it doesn't allow for the normal type inferences that `jsg::Unimplemented` does to generate it's error messages

Here is an idea I had for an alternative:
We have a templated type called `jsg::WasUnimplemented<T>`, then this type is also castable to `jsg::Unimplemented`.
In the `WasUnimplemented` constructor, the desired function call to check the feature flag is passed in, and immediately checked, this then triggers the same `jsg::Unimplemented` error.
I'm not 100% sure of all the details, but the essence is we need a type that represents something that we have now implemented and at one point hadn't to preserve our error messages.

Reply:
I don't understand. Why do you need the type inference? You can just copy exactly the error string that was generated before, for this specific field, and have your code in `validate()` throw that message -- it should then be identical behavior, right?

Reply:
My idea seems incoherent and doesn't makes sense, the validate has been added in the last commit.